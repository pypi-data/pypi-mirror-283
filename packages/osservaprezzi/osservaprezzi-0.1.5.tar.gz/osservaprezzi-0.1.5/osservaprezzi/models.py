"""Models module."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from datetime import datetime


@dataclass(frozen=True)
class GPSCoordinates:
    """GPS coordinates definition."""

    latitude: float
    longitude: float

    def to_json(self) -> dict[str, float]:
        """Return JSON representation of the object."""
        return {
            "lat": self.latitude,
            "lng": self.longitude,
        }


@dataclass(frozen=True)
class Fuel:
    """Fuel definition."""

    id: int
    price: float
    name: str
    fuel_id: int
    is_self: bool
    insert_date: datetime | None
    validity_date: datetime | None


@dataclass(frozen=True)
# pylint: disable-next=too-many-instance-attributes
class Station:
    """Station definition."""

    id: int
    name: str
    description: str
    fuels: list[Fuel]
    location: GPSCoordinates
    insert_date: datetime | None
    address: str | None
    brand: str | None
    email: str | None
    website: str | None
    phone: str | None
    company: str | None
    services: list[Service] | None


@dataclass(frozen=True)
class Marker:
    """Marker definition."""

    type: str
    extension: str
    content: str


@dataclass(frozen=True)
class Brand:
    """Brand definition."""

    id: int
    name: str
    markers: list[Marker]


@dataclass(frozen=True)
class Service:
    """Service definition."""

    id: str | int
    description: str
