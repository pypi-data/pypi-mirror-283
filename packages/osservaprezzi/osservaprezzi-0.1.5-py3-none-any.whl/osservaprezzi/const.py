"""Constants module."""

from __future__ import annotations

DEFAULT_TIMEOUT = 10
URL_API_ENDPOINT = "https://carburanti.mise.gov.it/ospzApi"
PATH_API_BRANDS = "registry/alllogos"
PATH_API_FUELS = "registry/fuels"
PATH_API_ZONES = "search/zone"
PATH_API_SERVICE_AREA = "registry/servicearea/%d"
REQUEST_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:128.0) Gecko/20100101 Firefox/128.0",
    "Origin": "https://carburanti.mise.gov.it",
    "Host": "carburanti.mise.gov.it",
}
