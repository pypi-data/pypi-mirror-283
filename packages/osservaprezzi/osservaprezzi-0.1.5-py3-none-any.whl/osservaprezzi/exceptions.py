"""Osservaprezzi client exception module."""

from __future__ import annotations


class OsservaprezziError(Exception):
    """Osservaprezzi error."""


class ApiError(OsservaprezziError):
    """Api error."""


class ApiTimeoutError(OsservaprezziError):
    """Api timeout error."""
