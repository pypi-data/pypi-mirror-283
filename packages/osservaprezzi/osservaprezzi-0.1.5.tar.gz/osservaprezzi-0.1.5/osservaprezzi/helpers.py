"""Helpers definition."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from .models import Brand, Fuel, GPSCoordinates, Marker, Service, Station


def brand_from_json(json: dict[str, Any]) -> Brand:
    """Create a brand from JSON."""
    return Brand(
        id=json.get("bandieraId", ""),
        name=json.get("bandiera", ""),
        markers=[marker_from_json(marker) for marker in json.get("logoMarkerList", [])],
    )


def fuel_from_json(json: dict[str, Any]) -> Fuel:
    """Create a fuel from JSON."""
    return Fuel(
        id=json.get("id", 0),
        price=json.get("price", 0),
        name=json.get("name", json.get("description", "")),
        fuel_id=json.get("fuelId", 0),
        is_self=json.get("isSelf", False),
        insert_date=datetime.fromisoformat(json["insertDate"])
        if "insertDate" in json
        else None,
        validity_date=datetime.fromisoformat(json["validityDate"])
        if "validityDate" in json
        else None,
    )


def gps_coordinates_from_json(json: dict[str, Any]) -> GPSCoordinates:
    """Create a GPS Coordinates from JSON."""
    return GPSCoordinates(
        latitude=json.get("lat", 0),
        longitude=json.get("lng", 0),
    )


def marker_from_json(json: dict[str, Any]) -> Marker:
    """Create a marker from JSON."""
    return Marker(
        type=json.get("tipoFile", ""),
        extension=json.get("estensione", ""),
        content=json.get("content", ""),
    )


def service_from_json(json: dict[str, Any]) -> Service:
    """Create a service from JSON."""
    return Service(
        id=json.get("id", ""),
        description=json.get("description", ""),
    )


def station_from_json(json: dict[str, Any]) -> Station:
    """Create a station from JSON."""
    return Station(
        id=json.get("id", 0),
        name=json.get("name", ""),
        description=json.get("nomeImpianto", ""),
        fuels=[fuel_from_json(fuel) for fuel in json.get("fuels", [])],
        location=gps_coordinates_from_json(json.get("location", {})),
        insert_date=datetime.fromisoformat(json["insertDate"])
        if "insertDate" in json
        else None,
        address=json.get("address"),
        brand=json.get("brand"),
        email=json.get("email"),
        website=json.get("website"),
        phone=json.get("phoneNumber"),
        company=json.get("company"),
        services=[service_from_json(service) for service in json.get("services", [])],
    )
