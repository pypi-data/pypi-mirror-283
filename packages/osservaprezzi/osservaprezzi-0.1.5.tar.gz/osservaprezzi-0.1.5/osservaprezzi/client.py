"""Osservaprezzi module."""

from __future__ import annotations

from http import HTTPStatus
import logging
from typing import TYPE_CHECKING, Any

from aiohttp import ClientError, ClientSession

from osservaprezzi.helpers import (
    brand_from_json,
    fuel_from_json,
    station_from_json,
)

from .const import (
    DEFAULT_TIMEOUT,
    PATH_API_BRANDS,
    PATH_API_FUELS,
    PATH_API_SERVICE_AREA,
    PATH_API_ZONES,
    REQUEST_HEADERS,
    URL_API_ENDPOINT,
)
from .exceptions import ApiError, ApiTimeoutError

if TYPE_CHECKING:
    from .models import Brand, Fuel, GPSCoordinates, Station

_LOGGER = logging.getLogger(__name__)


class Osservaprezzi:
    """Osservaprezzi client class."""

    def __init__(self, session: ClientSession) -> None:
        self._session = session

    async def get_brands(self) -> list[Brand]:
        """Get brands."""
        json = await self._get(PATH_API_BRANDS)
        return [brand_from_json(data) for data in json.get("loghi", [])]

    async def get_fuels(self) -> list[Fuel]:
        """Get registered fuels."""
        json = await self._get(PATH_API_FUELS)
        return [fuel_from_json(data) for data in json.get("results", [])]

    async def get_stations(
        self,
        coordinates: GPSCoordinates,
        *,
        radius: int = 5,
        fuel_type: str | None = None,
    ) -> list[Station]:
        """Get stations."""
        payload = {"points": [coordinates.to_json()], "radius": radius}
        if fuel_type is not None:
            payload["fuelType"] = fuel_type
        json = await self._post(PATH_API_ZONES, payload)
        return [station_from_json(data) for data in json.get("results", [])]

    async def get_station(self, station_id: int) -> Station | None:
        """Get station."""
        json = await self._get(PATH_API_SERVICE_AREA % station_id)
        return station_from_json(json) if "id" in json else None

    async def _get(self, path: str) -> dict[str, Any]:
        """Perform a GET request."""
        try:
            url = f"{URL_API_ENDPOINT}/{path}"
            async with self._session.get(
                url, timeout=DEFAULT_TIMEOUT, headers=REQUEST_HEADERS
            ) as response:
                if response.status == HTTPStatus.OK:
                    response_data: dict[str, Any] = await response.json()
                    return response_data
        except TimeoutError as ex:
            raise ApiTimeoutError from ex
        except ClientError as ex:
            raise ApiError from ex

        raise ApiError("Unknown error occurred")

    async def _post(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        """Perform a POST request."""
        try:
            url = f"{URL_API_ENDPOINT}/{path}"
            async with self._session.post(
                url,
                json=payload,
                timeout=DEFAULT_TIMEOUT,
                headers=REQUEST_HEADERS,
            ) as response:
                if response.status == HTTPStatus.OK:
                    response_data: dict[str, Any] = await response.json()
                    return response_data
        except TimeoutError as ex:
            raise ApiTimeoutError from ex
        except ClientError as ex:
            raise ApiError from ex

        raise ApiError("Unknown error occurred")
