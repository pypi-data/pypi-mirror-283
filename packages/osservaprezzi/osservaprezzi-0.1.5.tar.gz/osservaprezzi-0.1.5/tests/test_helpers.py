from __future__ import annotations

from typing import TYPE_CHECKING, Any

import pytest

from osservaprezzi.helpers import (
    brand_from_json,
    fuel_from_json,
    gps_coordinates_from_json,
    marker_from_json,
    service_from_json,
    station_from_json,
)

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion


@pytest.mark.parametrize(
    "json",
    [
        {"bandieraId": 1, "bandiera": "Brand #1"},
        {
            "bandieraId": 2,
            "bandiera": "Brand #2",
            "logoMarkerList": [
                {
                    "tipoFile": "logo",
                    "estensione": "png",
                    "content": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
                }
            ],
        },
    ],
)
async def test_brand_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="brand_from_json") == brand_from_json(json)


@pytest.mark.parametrize(
    "json",
    [
        {"id": "1-x", "description": "Benzina"},
        {
            "id": 75557219,
            "price": 2.079,
            "name": "Benzina",
            "fuelId": 1,
            "isSelf": False,
        },
        {
            "id": 75557219,
            "price": 2.079,
            "name": "Benzina",
            "fuelId": 1,
            "isSelf": False,
            "insertDate": "2024-07-02T12:00:38Z",
            "validityDate": "2024-07-02T11:59:00Z",
        },
    ],
)
async def test_fuel_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="fuel_from_json") == fuel_from_json(json)


@pytest.mark.parametrize(
    "json",
    [{"lat": 0, "lng": 0}, {"lat": 45.541553, "lng": 10.211802}],
)
async def test_gps_coordinates_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="gps_coordinates_from_json") == gps_coordinates_from_json(json)


@pytest.mark.parametrize(
    "json",
    [
        {
            "tipoFile": "logo",
            "estensione": "png",
            "content": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==",
        }
    ],
)
async def test_marker_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="marker_from_json") == marker_from_json(json)


@pytest.mark.parametrize(
    "json",
    [{"id": 1, "description": "Dummy Service"}, {"id": 6, "description": "Bancomat"}],
)
async def test_service_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="service_from_json") == service_from_json(json)


@pytest.mark.parametrize(
    "json",
    [
        {
            "id": 1,
            "name": "Service Area 1",
            "fuels": [
                {
                    "id": 10001,
                    "price": 2.079,
                    "name": "Benzina",
                    "fuelId": 1,
                    "isSelf": False,
                },
                {
                    "id": 10002,
                    "price": 1.869,
                    "name": "Benzina",
                    "fuelId": 1,
                    "isSelf": True,
                },
                {
                    "id": 10003,
                    "price": 1.989,
                    "name": "Gasolio",
                    "fuelId": 2,
                    "isSelf": False,
                },
                {
                    "id": 10004,
                    "price": 1.779,
                    "name": "Gasolio",
                    "fuelId": 2,
                    "isSelf": True,
                },
            ],
            "location": {
                "lat": 1.5,
                "lng": 5.9,
            },
            "insertDate": "1985-05-01T18:20:00+01:00",
            "address": "This is the service area address",
            "brand": "Octan",
        },
        {
            "id": 2,
            "name": "Service Area 2",
            "fuels": [
                {
                    "id": 20001,
                    "price": 1.839,
                    "name": "Benzina",
                    "fuelId": 1,
                    "isSelf": True,
                },
                {
                    "id": 20002,
                    "price": 1.709,
                    "name": "Gasolio",
                    "fuelId": 2,
                    "isSelf": True,
                },
            ],
            "location": {"lat": 1.6, "lng": 4.7},
            "insertDate": "1985-09-05T11:56:40+01:00",
            "brand": "Octan",
        },
        {
            "id": 1,
            "name": "Octan Service Area",
            "nomeImpianto": "Octan Service Area",
            "address": "Service Area Address",
            "brand": "Octan",
            "fuels": [
                {
                    "id": 20001,
                    "price": 1.839,
                    "name": "Benzina",
                    "fuelId": 1,
                    "isSelf": True,
                    "serviceAreaId": 1,
                    "insertDate": "2024-07-02T12:00:38Z",
                    "validityDate": "2024-07-02T11:59:00Z",
                },
                {
                    "id": 20002,
                    "price": 1.709,
                    "name": "Gasolio",
                    "fuelId": 2,
                    "isSelf": True,
                    "serviceAreaId": 1,
                    "insertDate": "2024-07-02T12:00:38Z",
                    "validityDate": "2024-07-02T11:59:00Z",
                },
            ],
            "phoneNumber": "",
            "email": "",
            "website": "",
            "company": "Ugo Legozzi",
            "services": [{"id": 6, "description": "Bancomat"}],
        },
    ],
)
async def test_station_from_json(
    snapshot: SnapshotAssertion,
    json: dict[str, Any],
) -> None:
    assert snapshot(name="station_from_json") == station_from_json(json)
