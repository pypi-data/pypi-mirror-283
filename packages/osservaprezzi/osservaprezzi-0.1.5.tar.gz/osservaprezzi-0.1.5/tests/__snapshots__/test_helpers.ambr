# serializer version: 1
# name: test_brand_from_json[json0][brand_from_json]
  Brand(id=1, name='Brand #1', markers=[])
# ---
# name: test_brand_from_json[json1][brand_from_json]
  Brand(id=2, name='Brand #2', markers=[Marker(type='logo', extension='png', content='iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==')])
# ---
# name: test_fuel_from_json[json0][fuel_from_json]
  Fuel(id='1-x', price=0, name='Benzina', fuel_id=0, is_self=False, insert_date=None, validity_date=None)
# ---
# name: test_fuel_from_json[json1][fuel_from_json]
  Fuel(id=75557219, price=2.079, name='Benzina', fuel_id=1, is_self=False, insert_date=None, validity_date=None)
# ---
# name: test_fuel_from_json[json2][fuel_from_json]
  Fuel(id=75557219, price=2.079, name='Benzina', fuel_id=1, is_self=False, insert_date=datetime.datetime(2024, 7, 2, 12, 0, 38, tzinfo=datetime.timezone.utc), validity_date=datetime.datetime(2024, 7, 2, 11, 59, tzinfo=datetime.timezone.utc))
# ---
# name: test_gps_coordinates_from_json[json0][gps_coordinates_from_json]
  GPSCoordinates(latitude=0, longitude=0)
# ---
# name: test_gps_coordinates_from_json[json1][gps_coordinates_from_json]
  GPSCoordinates(latitude=45.541553, longitude=10.211802)
# ---
# name: test_marker_from_json[json0][marker_from_json]
  Marker(type='logo', extension='png', content='iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8BQDwAEhQGAhKmMIQAAAABJRU5ErkJggg==')
# ---
# name: test_service_from_json[json0][service_from_json]
  Service(id=1, description='Dummy Service')
# ---
# name: test_service_from_json[json1][service_from_json]
  Service(id=6, description='Bancomat')
# ---
# name: test_station_from_json[json0][station_from_json]
  Station(id=1, name='Service Area 1', description='', fuels=[Fuel(id=10001, price=2.079, name='Benzina', fuel_id=1, is_self=False, insert_date=None, validity_date=None), Fuel(id=10002, price=1.869, name='Benzina', fuel_id=1, is_self=True, insert_date=None, validity_date=None), Fuel(id=10003, price=1.989, name='Gasolio', fuel_id=2, is_self=False, insert_date=None, validity_date=None), Fuel(id=10004, price=1.779, name='Gasolio', fuel_id=2, is_self=True, insert_date=None, validity_date=None)], location=GPSCoordinates(latitude=1.5, longitude=5.9), insert_date=datetime.datetime(1985, 5, 1, 18, 20, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), address='This is the service area address', brand='Octan', email=None, website=None, phone=None, company=None, services=[])
# ---
# name: test_station_from_json[json1][station_from_json]
  Station(id=2, name='Service Area 2', description='', fuels=[Fuel(id=20001, price=1.839, name='Benzina', fuel_id=1, is_self=True, insert_date=None, validity_date=None), Fuel(id=20002, price=1.709, name='Gasolio', fuel_id=2, is_self=True, insert_date=None, validity_date=None)], location=GPSCoordinates(latitude=1.6, longitude=4.7), insert_date=datetime.datetime(1985, 9, 5, 11, 56, 40, tzinfo=datetime.timezone(datetime.timedelta(seconds=3600))), address=None, brand='Octan', email=None, website=None, phone=None, company=None, services=[])
# ---
# name: test_station_from_json[json2][station_from_json]
  Station(id=1, name='Octan Service Area', description='Octan Service Area', fuels=[Fuel(id=20001, price=1.839, name='Benzina', fuel_id=1, is_self=True, insert_date=datetime.datetime(2024, 7, 2, 12, 0, 38, tzinfo=datetime.timezone.utc), validity_date=datetime.datetime(2024, 7, 2, 11, 59, tzinfo=datetime.timezone.utc)), Fuel(id=20002, price=1.709, name='Gasolio', fuel_id=2, is_self=True, insert_date=datetime.datetime(2024, 7, 2, 12, 0, 38, tzinfo=datetime.timezone.utc), validity_date=datetime.datetime(2024, 7, 2, 11, 59, tzinfo=datetime.timezone.utc))], location=GPSCoordinates(latitude=0, longitude=0), insert_date=None, address='Service Area Address', brand='Octan', email='', website='', phone='', company='Ugo Legozzi', services=[Service(id=6, description='Bancomat')])
# ---
